Sources/Apollo/HTTPNetworkTransport.swift

```diff
@@ -1,11 +1,78 @@
+  ///
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport which is about to send a request
+  ///   - request: The request, as an `inout` variable for modification
+  func networkTransport(_ networkTransport: HTTPNetworkTransport, willSend request: inout URLRequest)
+}
+
+// MARK: -
+
+/// Methods which will be called after some kind of response has been received to a `URLSessionTask`.
+public protocol HTTPNetworkTransportTaskCompletedDelegate {
+  
+  /// A callback to allow hooking in URL session responses for things like logging and examining headers.
+  /// NOTE: This will call back on whatever thread the URL session calls back on, which is never the main thread. Call `DispatchQueue.main.async` before touching your UI!
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport that completed a task
+  ///   - request: The request which was completed by the task
+  ///   - data: [optional] Any data received. Passed through from `URLSession`.
+  ///   - response: [optional] Any response received. Passed through from `URLSession`.
+  ///   - error: [optional] Any error received. Passed through from `URLSession`.
+  func networkTransport(_ networkTransport: HTTPNetworkTransport,
+                        completedRawTaskForRequest request: URLRequest,
+                        withData data: Data?,
+                        response: URLResponse?,
+                        error: Error?)
+}
+
+// MARK: -
+
+public protocol HTTPNetworkTransportRetryDelegate: class {
+  
+  /// Called when an error has been received after a request has been sent to the server to see if an operation should be retried or not.
+  /// NOTE: Don't just call the `retryHandler` with `true` all the time, or you can potentially wind up in an infinite loop of errors
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport which received the error
+  ///   - error: The received error
+  ///   - request: The URLRequest which generated the error
+  ///   - response: [Optional] Any response received when the error was generated
+  ///   - retryHandler: A closure indicating whether the operation should be retried. Asyncrhonous to allow for re-authentication or other async operations to complete.
+  func networkTransport(_ networkTransport: HTTPNetworkTransport,
+                        receievedError error: Error,
+                        for request: URLRequest,
+                        response: URLResponse?,
+                        retryHandler: @escaping (_ shouldRetry: Bool) -> Void)
+}
+
+// MARK: -
+
 /// A network transport that uses HTTP POST requests to send GraphQL operations to a server, and that uses `URLSession` as the networking implementation.
 public class HTTPNetworkTransport: NetworkTransport {
   let url: URL
   let session: URLSession
   let serializationFormat = JSONSerializationFormat.self
   let useGETForQueries: Bool
+  let preflightDelegate: HTTPNetworkTransportPreflightDelegate?
```

Comment:
I'm a little worried about a proliferation of delegate properties here. I can see the benefit of keeping these separate, but it may also be worth simplifying this to a single `delegate` property that can conform to any of these delegate protocols (similar to how `URLSession`s delegate can conform to `URLSessionDelegate`, `URLSessionTaskDelegate`, etc.).

Reply:
The problem is that `URLSessionDelegate` et al are Objective-C protocols, so they can have optional methods. If you only want to implement method `foo()`, then that's all you have to implement.

Our Swift protocols can't unless we mark the protocol `@objc`, which has a ton of other knock-on effects we don't want. This means if you're implementing a single protocol, you have to implement ALL of its methods even if you don't need or even want them. 

I agree that having to have things conforming to multiple protocols is annoying, but I think the separation of concerns (and the ability to avoid needing to implement EVERYTHING just to get one thing) is worth it.

Reply:
Ah, I'm ok with having different delegate protocols, but what I meant was that we could still have only one `delegate` property. I believe that is similar to what `URLSession`'s delegate does, so your delegate has to conform to `URLSessionDelegate` and optionally to `URLSessionTaskDelegate` etc.

Reply:
`URLSessionDelegate` is the base protocol which inherits from `NSObjectProtocol` - `URLSessionTaskDelegate` et al inherit from `URLSessionDelegate. Since these are all Obj-C, they can have optional methods, which pure Swift protocols can't. 

We could set up some kind of inheritance with our pure Swift protocols to allow passing the highest level as a single delegate, but the user would still need to implement all protocols in the inheritance tree, which again may not be what they need or want.

Reply:
AHA. Talking through with Martijn I realized there's a way around this: [An empty base protocol](https://gist.github.com/designatednerd/a3d5033a43ad9ac1fb3125b4589a7cfc). Implementing!